docs/semgrep-code/java.md

```diff
@@ -0,0 +1,100 @@
+
+* This example has two true positives: **line 11** and **line 20**.
+* Semgrep Pro is able to detect that **line 14 and 17 are false positives**. Semgrep OSS can't catch that distinction.
+* Lines 14 and 17 are false positives because `input.endsWith("something")` and `input.indexOf('u')` return a boolean and integer respectively. Semgrep Pro is able to understand `endsWith` and `indexOf` Java methods.
+* The Semgrep rule uses the fields `taint_assume_safe_booleans` and `taint_assume_safe_numbers` to tell the engine that these types are safe and not tainted. 
+
+## Target code in a parent class and its subclasses
+
+Semgrep supports class inheritance in Java. You can use Semgrep to search across all subclasses. This specificity means that rules can better target your codebase, increasing true positive rates. This is achieved through the `metavariable-type` field, which can accept the name of any user-defined class. 
+
+The `metavariable-type` field is available in Semgrep OSS. However, classes in Java are frequently defined across files (interfile), which is beyond the scope of Semgrep OSS's analysis. Use Semgrep Pro to perform cross-file analysis to ensure that Semgrep can detect all class and subclass definitions.
+
+### Example: `detect-pattern-in-subclass`
+
+<iframe title="detect-pattern-in-subclass" src="https://semgrep.dev/embed/editor?snippet=nJjjG" width="100%" height="432px" frameBorder="0"></iframe>
+
+**Figure**. `detect-pattern-in-subclass`. To view the entire sample code and rule, click **Open in Playground**.
+
+This demo rule detects patterns in instances of the user-defined parent class `Foo` and its subclasses.
+
+- This example has two true positives: **line 10** and **line 25**.
+- The `patterns` array initially defines a `pattern: $CLASS.x`.
+    - **Line 17**, `baz.x` fulfills this pattern.
+    - However, the `metavariable-type` specifies a `type` of `Foo`.
+    - This specification narrows the match to **line 10** because `Bar` is a subclass of `Foo`, and **line 25**, which is an instance of the Foo object itself.
+
+## Semgrep supports field and index sensitivity
+
+Field sensitivity means that Semgrep can track taint for each field of an object, independently. Given an object `C` with properties `C.x` and `C.y`, if `C.x` is tainted, then it does **not** automatically mark `C.y` as tainted.
+
+Similarly, index sensitivity means that Semgrep can track taint for each element of an array, independently.
+
+### Example: `unsafe-sql-concatenation-in-method-taint-field-sensitivity`
+
+<iframe title="unsafe-sql-concatenation-in-method-taint-field-sensitivity" src="https://semgrep.dev/embed/editor?snippet=OrAwe" width="100%" height="432px" frameBorder="0"></iframe>
+
+**Figure**. `unsafe-sql-concatenation-in-method-taint-field-sensitivity`. To view the entire sample code and rule, click **Open in Playground**.
```

Comment:
Does something like "Figure: `unsafe-sql-concatenation-in-method-taint-field-sensitivity` ..." work?

Reply:
It's been shortened, fear not.